lib/kauth/login.go

```diff
@@ -0,0 +1,121 @@
+package kauth
+
+import (
+	"context"
+	"crypto/ed25519"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"github.com/enfabrica/enkit/astore/common"
+	"github.com/enfabrica/enkit/astore/rpc/auth"
+	"github.com/enfabrica/enkit/lib/kcerts"
+	"github.com/enfabrica/enkit/lib/logger"
+	"github.com/enfabrica/enkit/lib/retry"
+	"github.com/enfabrica/enkit/lib/srand"
+	"github.com/pkg/browser"
+	"golang.org/x/crypto/nacl/box"
+	"golang.org/x/crypto/ssh"
+	"math/rand"
+)
+
+func init() {
+	browser.Stdout = nil
+	browser.Stderr = nil
+}
+
+type EnkitCredentials struct {
+	Token string
+	// The below fields can be possibly empty if the auth server does not support CA certificates.
+	CaHosts        []string
+	CAPublicKey    string
+	PrivateKey     ed25519.PrivateKey
+	SSHCertificate *ssh.Certificate
+}
+
+// PerformLogin will login with the provider auth client, retry and logger. It does not care about the cache.
+// If you wish to save the result, please call SaveCredentials
+func PerformLogin(authClient auth.AuthClient, l logger.Logger, repeater *retry.Options, username, domain string) (*EnkitCredentials, error) {
+	rng := rand.New(srand.Source)
+	pubBox, privBox, err := box.GenerateKey(rng)
+	if err != nil {
+		return nil, err
+	}
+	sshPub, sshPriv, err := kcerts.MakeKeys(kcerts.GenerateED25519)
+	if err != nil {
+		return nil, err
+	}
+	areq := &auth.AuthenticateRequest{
+		Key:    (*pubBox)[:],
+		User:   username,
+		Domain: domain,
+	}
+	l.Infof("Retrieving authentication url.")
+	ares, err := authClient.Authenticate(context.TODO(), areq)
+	if err != nil {
+		return nil, fmt.Errorf("Could not contact the authentication server. Is your connectivity working? Is the server up?\nFor debugging: %w", err)
+	}
+	if username != "" {
+		fmt.Printf("Dear %s, please visit:\n\n", username)
+	} else {
+		fmt.Printf("Kind human, please visit:\n\n")
+	}
+	fmt.Printf("\t%s\n\nTo complete authentication with @%s.\n"+
+		"I'll be waiting for you, but hurry! The request may timeout.\nHit Ctl+C with no regrets to abort.\n", ares.Url, domain)
+	l.Infof("Authentication url is %s, attempting to open with your Os's default browser", ares.Url)
+	// browser.OpenURL blocks depending on permissions level and OS. By running it in a goroutine, we ensure that
+	// the login process does not get stuck waiting for the browser window to be closed.
+	go func() {
```

Comment:
This PR is not just a refactor! It is solving a bug! Bugfixes should go in THEIR OWN PR.
That way, we can cherrypick, or if we rollback your larger refactor due to a separate bug, we don't lose the dedicate bugfix.

This is pretty minor, but please make sure your merge COMMIT MESSAGE includes the fact that there is a bugfix here. From `git log`, my assumption for this PR is that this should be a noop in terms of behavior.

Reply:
Updated the title so that when it gets squashed it'll show up.